components/ws-proxy/pkg/sshproxy/forward.go

```diff
@@ -16,178 +14,79 @@ import (
-	closer.Do(closeFunc)
-}
-
-func (s *Server) SessionForward(session *Session, newChannel ssh.NewChannel) {
-	sessChan, sessReqs, err := newChannel.Accept()
-	if err != nil {
-		log.WithError(err).Error("SessionForward accept failed")
-		return
-	}
-	defer sessChan.Close()
+	defer clientChan.Close()
 
 	maskedReqs := make(chan *ssh.Request, 1)
+
 	go func() {
-		for req := range sessReqs {
+		for req := range clientReqs {
 			switch req.Type {
 			case "pty-req", "shell":
 				log.WithFields(log.OWI("", session.WorkspaceID, session.InstanceID)).Debugf("forwarding %s request", req.Type)
 				if req.WantReply {
 					req.Reply(true, []byte{})
 					req.WantReply = false
 				}
-			case "keepalive@openssh.com":
-				if req.WantReply {
-					req.Reply(true, []byte{})
-					req.WantReply = false
-				}
 			}
 			maskedReqs <- req
 		}
+		close(maskedReqs)
 	}()
 
-	stderr := sessChan.Stderr()
-
-	clientConfig := &ssh.ClientConfig{
-		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
-		User:            GitpodUsername,
-		Auth: []ssh.AuthMethod{
-			ssh.PublicKeysCallback(func() (signers []ssh.Signer, err error) {
-				return []ssh.Signer{session.WorkspacePrivateKey}, nil
-			}),
-		},
-		Timeout: 10 * time.Second,
-	}
-
-	if s.ConnectionTimeout != 0 {
-		clientConfig.Timeout = s.ConnectionTimeout
-	}
+	go func() {
+		io.Copy(workspaceChan, clientChan)
+		workspaceChan.CloseWrite()
+	}()
 
-	conn, err := net.Dial("tcp", session.WorkspaceIP)
-	if err != nil {
-		fmt.Fprintf(stderr, "Connect failed: %v\r\n", err)
-		return
-	}
-	defer conn.Close()
+	go func() {
+		io.Copy(clientChan, workspaceChan)
+		clientChan.CloseWrite()
+	}()
 
-	clientConn, clientChans, clientReqs, err := ssh.NewClientConn(conn, session.WorkspaceIP, clientConfig)
-	if err != nil {
-		fmt.Fprintf(stderr, "Client connection setup failed: %v\r\n", err)
-		return
+	wg := sync.WaitGroup{}
+	forward := func(sourceReqs <-chan *ssh.Request, targetChan ssh.Channel) {
+		defer wg.Done()
+		for ctx.Err() == nil {
+			select {
+			case req, ok := <-sourceReqs:
+				if !ok {
+					targetChan.Close()
+					return
+				}
+				b, err := targetChan.SendRequest(req.Type, req.WantReply, req.Payload)
+				_ = req.Reply(b, nil)
+				if err != nil {
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
 	}
-	client := ssh.NewClient(clientConn, clientChans, clientReqs)
 
-	forwardChannel, forwardReqs, err := client.OpenChannel("session", []byte{})
-	if err != nil {
-		fmt.Fprintf(stderr, "Remote session setup failed: %v\r\n", err)
-		return
-	}
+	wg.Add(2)
+	go forward(maskedReqs, workspaceChan)
+	go forward(workspaceReqs, clientChan)
 
-	proxy(maskedReqs, forwardReqs, startHeartbeatingChannel(sessChan, s.Heartbeater, session.InstanceID), forwardChannel)
+	wg.Wait()
+	log.WithFields(log.OWI("", session.WorkspaceID, session.InstanceID)).Debug("session forward stop")
 }
 
 func startHeartbeatingChannel(c ssh.Channel, heartbeat Heartbeat, instanceID string) ssh.Channel {
```

Comment:
uh, i overlooked it before, do we already have heartbeating? Does it only work on pty input? We don't want to keep workspaces alive just for established ssh connection or some port forwarding.